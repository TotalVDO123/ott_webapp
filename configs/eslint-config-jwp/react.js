module.exports = {
  extends: [
    // Extend the base config
    './typescript',

    // Use recommended React rules
    'plugin:react/recommended',
  ],

  rules: {
    // Not needed in React 17
    'react/react-in-jsx-scope': 'off',
  },

  overrides: [
    {
      files: ['*.jsx', '*.tsx', '*.ts'],
      plugins: [
        // Enable linting React code
        'react',
        'react-hooks',
      ],
      rules: {
        // Help with Hooks syntax
        'react-hooks/rules-of-hooks': 'error',
        'react-hooks/exhaustive-deps': 'error',

        // Handled by Typescript
        'react/prop-types': 'off',

        // This rule causes too many false positives, e.g. with default exports or child render function
        'react/display-name': 'off',
      },
    },
  ],

  settings: {
    react: {
      pragma: 'React',
      version: '17',
    },
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
  },
};
