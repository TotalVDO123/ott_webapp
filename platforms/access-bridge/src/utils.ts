import { IncomingMessage } from 'node:http';

export function isValidSiteId(siteId: string): boolean {
  // Regular expression to match exactly 8 alphanumeric characters
  const alphanumericRegex = /^[a-zA-Z0-9]{8}$/;
  return alphanumericRegex.test(siteId);
}

// parse request body provided in an API call
export const parseJsonBody = <T>(req: IncomingMessage): Promise<T> => {
  return new Promise((resolve, reject) => {
    let body = '';

    req.on('data', (chunk) => {
      body += chunk;
    });

    req.on('end', () => {
      try {
        resolve(JSON.parse(body) as T);
      } catch (error) {
        reject(new Error('Invalid JSON'));
      }
    });

    req.on('error', (error) => {
      reject(error);
    });
  });
};
