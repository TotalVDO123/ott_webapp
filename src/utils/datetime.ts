/**
 * Returns true when the given string is a valid date string
 */
export function isValidDateString(input: string) {
  return input ? new Date(input).toString() !== 'Invalid Date' : false;
}

/**
 * Seconds to ISO8601 duration or date string
 */
export function secondsToISO8601(input: number, timeOnly: boolean = false): string {
  if (!input) {
    return '';
  }

  const date = new Date(input ? input * 1000 : 0);
  const hours = date.getUTCHours();
  const minutes = date.getUTCMinutes();
  const seconds = date.getUTCSeconds();

  if (!timeOnly) {
    return date.toISOString();
  }

  let isoString = 'PT';
  if (hours > 0) isoString += hours + 'H';
  if (minutes > 0) isoString += minutes + 'M';
  if (seconds > 0) isoString += seconds + 'S';

  return isoString;
}

/**
 * Returns a Date instance with the time set to the beginning of the day
 */
export const startOfDay = () => {
  const date = new Date();
  date.setHours(0, 0, 0);

  return date;
};

/**
 * Returns a Date instance with the time set to the end of the day
 */
export const endOfDay = () => {
  const date = new Date();
  date.setHours(23, 59, 59);

  return date;
};
