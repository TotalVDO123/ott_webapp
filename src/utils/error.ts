export class SettingsError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SettingsError';
  }
}

export class ConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ConfigError';
  }
}
