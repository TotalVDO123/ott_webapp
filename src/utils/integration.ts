import { INTEGRATION } from '#src/config';
import type { Config } from '#types/Config';

interface IntegrationInfo {
  integrationType: keyof typeof INTEGRATION | null;
  useSandbox?: boolean;
  clientId?: string;
  offers?: string[];
}
export default function getIntegration(integrations: Config['integrations']): IntegrationInfo {
  if (integrations?.cleeng?.id && integrations?.jwp?.clientId) {
    throw new Error('Invalid client integration. You cannot have both Cleeng and JWP integrations enabled at the same time.');
  }

  if (integrations?.jwp?.clientId) {
    return {
      integrationType: INTEGRATION.JWP,
      useSandbox: !!integrations.jwp.useSandbox,
      clientId: integrations.jwp.clientId,
      offers: integrations.jwp.assetId ? [`${integrations.jwp.assetId}`] : [],
    };
  }

  if (integrations?.cleeng?.id) {
    return {
      integrationType: INTEGRATION.CLEENG,
      useSandbox: !!integrations.cleeng.useSandbox,
      clientId: integrations.cleeng.id,
      offers: [integrations.cleeng?.monthlyOffer, integrations.cleeng?.yearlyOffer].filter((offer) => !!offer).map((offer) => `${offer}`),
    };
  }

  return {
    integrationType: null,
  };
}
