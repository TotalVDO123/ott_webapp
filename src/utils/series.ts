import type { EpisodeMetadata, Series } from '#types/series';
import { getEpisodes, getSeasonWithEpisodes } from '#src/services/api.service';

/**
 * Get an array of options for a season filter
 */
export const getFiltersFromSeries = (series: Series | undefined): { label: string; value: string }[] => {
  return series && 'seasons' in series
    ? series.seasons.map((season) => ({ label: season.season_title || String(season.season_number), value: String(season.season_number) }))
    : [];
};

/**
 * Retrieve a first episode of the next season
 */
const getNextSeasonEpisode = async (seasonNumber: number, series: Series | undefined) => {
  const seasons = series?.seasons;

  // The seasons is the last one in case there is only one season available or it is the same as the last season in the seasons array
  const isLastSeason = seasons?.length === 1 || seasons?.[seasons?.length - 1]?.season_number === seasonNumber;

  if (isLastSeason) {
    return;
  } else {
    // Need to know the order of seasons to choose the next season number
    const hasAscendingOrder = (Number(seasons?.[1]?.season_number) || 0) > (Number(seasons?.[0]?.season_number) || 0);
    const nextSeasonNumber = hasAscendingOrder ? seasonNumber + 1 : seasonNumber - 1;

    return (await getSeasonWithEpisodes(series?.series_id, nextSeasonNumber, 0, 1))?.episodes?.[0];
  }
};

/**
 * Get a next episode of the selected season / episodes list
 */
const getNextEpisode = async (seasonNumber: number, seriesId: string, pageWithEpisode: number) => {
  if (seasonNumber) {
    return (await getSeasonWithEpisodes(seriesId, seasonNumber, pageWithEpisode, 1))?.episodes?.[0];
  } else {
    return (await getEpisodes(seriesId, pageWithEpisode, 1))?.episodes?.[0];
  }
};

export const getNextItem = async (series: Series | undefined, episodeMetadata: EpisodeMetadata | undefined) => {
  if (!episodeMetadata || !series) {
    return;
  }

  const seasonNumber = Number(episodeMetadata.seasonNumber);
  const episodeNumber = Number(episodeMetadata.episodeNumber);

  // Get initial data to collect information about total number of elements
  const { episodes, pagination } =
    seasonNumber !== 0 ? await getSeasonWithEpisodes(series?.series_id, seasonNumber, 0) : await getEpisodes(series?.series_id, 0);

  if (episodes.length === 1 && seasonNumber) {
    return getNextSeasonEpisode(seasonNumber, series);
  }

  // Both episodes and seasons can be sorted by asc / desc
  const hasAscendingOrder = (Number(episodes[1].episodeNumber) || 0) > (Number(episodes[0].episodeNumber) || 0);
  const nextEpisodeNumber = hasAscendingOrder ? episodeNumber + 1 : episodeNumber - 1;
  // First page has 0 number, that is why we use Math.floor here und subtract 1
  const pageWithEpisode = Math.floor(hasAscendingOrder ? nextEpisodeNumber - 1 : pagination.total - nextEpisodeNumber - 1);
  // Consider the case when we have a next episode in the retrieved list
  const nextElementIndex = episodes.findIndex((el) => Number(el.episodeNumber) === nextEpisodeNumber);

  if (nextElementIndex !== -1) {
    return episodes[nextElementIndex];
  }

  // Fetch the next episodes of the season when there is more to fetch
  if (pageWithEpisode < pagination.total) {
    return getNextEpisode(seasonNumber, series?.series_id, pageWithEpisode);
    // Switch selected season in case the current one has nor more episodes inside
  } else if (seasonNumber) {
    return getNextSeasonEpisode(seasonNumber, series);
  }
};

/** Get a total amount of episodes in a season */
export const getEpisodesInSeason = (episodeMetadata: EpisodeMetadata | undefined, series: Series | undefined) => {
  return (series?.seasons || []).find((el) => el.season_number === Number(episodeMetadata?.seasonNumber))?.episode_count;
};
