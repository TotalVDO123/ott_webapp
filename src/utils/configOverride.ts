// Use session storage so the override persists until the tab is closed and then resets
import { logDev } from '#src/utils/common';

const storage = window.sessionStorage;

export const configQueryKey = 'app-config';
const configLegacyQueryKey = 'c';

const configFileStorageKey = 'config-file-override';

const DEFAULT_CONFIG_SOURCE = import.meta.env.APP_CONFIG_DEFAULT_SOURCE;
const ALLOWED_SOURCES = import.meta.env.APP_CONFIG_ALLOWED_SOURCES?.split(' ').filter((c) => c.toLowerCase() !== DEFAULT_CONFIG_SOURCE?.toLowerCase()) || [];
const UNSAFE_ALLOW_DYNAMIC_CONFIG = ['1', 'true'].includes(import.meta.env.APP_UNSAFE_ALLOW_DYNAMIC_CONFIG?.toLowerCase() || '');

export function getConfigLocation() {
  // Require a default source unless the dynamic (demo) mode is enabled
  if (!DEFAULT_CONFIG_SOURCE && !UNSAFE_ALLOW_DYNAMIC_CONFIG) {
    throw 'A default config is required';
  }

  return getConfigOverride() || DEFAULT_CONFIG_SOURCE;
}

export function addConfigQueryParam(config?: string) {
  const selectedConfig = config || getConfigLocation();

  // Make sure the config location is appended to the url,
  // but only when dynamic (demo) mode is enabled or using multiple configs and not the default
  if (selectedConfig && (UNSAFE_ALLOW_DYNAMIC_CONFIG || selectedConfig !== DEFAULT_CONFIG_SOURCE)) {
    const url = new URL(window.location.href);

    if (url.searchParams.get(configQueryKey) !== selectedConfig) {
      url.searchParams.set(configQueryKey, selectedConfig);

      window.history.replaceState(null, '', url.toString());
    }
  }
}

const getStoredConfig = () => {
  return storage.getItem(configFileStorageKey);
};

export const clearStoredConfig = () => {
  storage.removeItem(configFileStorageKey);

  const url = new URL(window.location.href);

  url.searchParams.delete(configQueryKey);
  window.history.replaceState(null, '', url.toString());
};

function getConfigOverride() {
  if (!UNSAFE_ALLOW_DYNAMIC_CONFIG && ALLOWED_SOURCES.length <= 0) {
    return undefined;
  }

  const url = new URL(window.location.href);

  // If the query string has the legacy key, remove it
  if (url.searchParams.has(configLegacyQueryKey)) {
    const legacyValue = url.searchParams.get(configLegacyQueryKey);
    url.searchParams.delete(configLegacyQueryKey);

    // If the new query key is not set, set it from the old key, but do not replace it (the new key 'wins' if both are set)
    if (legacyValue !== null && !url.searchParams.has(configQueryKey)) {
      url.searchParams.set(configQueryKey, legacyValue);
    }

    window.history.replaceState(null, '', url.toString());
  }

  if (url.searchParams.has(configQueryKey)) {
    const configQuery = url.searchParams.get(configQueryKey);

    // If the query param exists but the value is empty, clear the storage and allow fallback to the default config
    if (!configQuery) {
      clearStoredConfig();
      return undefined;
    }

    // If it's valid, store it and return it
    if (isValidConfigSource(configQuery)) {
      storage.setItem(configFileStorageKey, configQuery);
      return configQuery;
    }

    logDev(`Invalid app-config: ${configQuery}`);

    // Remove the query param if it's invalid
    url.searchParams.delete(configQueryKey);
    window.history.replaceState(null, '', url.toString());

    // Yes this falls through to look up the stored value if the query string is invalid and that's OK
  }

  const storedSource = getStoredConfig();

  // Make sure the stored value is still valid before returning it
  if (storedSource && isValidConfigSource(storedSource)) {
    return storedSource;
  }

  return undefined;
}

function isValidConfigSource(source: string) {
  // Dynamic values are valid as long as they are defined
  if (UNSAFE_ALLOW_DYNAMIC_CONFIG) {
    return !!source;
  }

  return DEFAULT_CONFIG_SOURCE === source || ALLOWED_SOURCES.indexOf(source) >= 0;
}
