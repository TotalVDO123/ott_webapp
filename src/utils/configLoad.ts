import merge from 'lodash.merge';

import { calculateContrastColor } from '../utils/common';
import { getConfigLocation } from '../utils/configOverride';
import loadConfig, { validateConfig } from '../services/config.service';
import { addScript } from '../utils/dom';
import { useConfigStore } from '../stores/ConfigStore';

import type { AccessModel, Config, Styling } from '#types/Config';

const CONFIG_HOST = import.meta.env.APP_API_BASE_URL;

const defaultConfig: Config = {
  id: '',
  siteName: '',
  description: '',
  player: '',
  assets: {
    banner: '/images/logo.png',
  },
  content: [],
  menu: [],
  integrations: {
    cleeng: {
      id: null,
      useSandbox: true,
    },
  },
  styling: {
    footerText: '',
    shelfTitles: true,
  },
  features: {
    enableSharing: true,
  },
};

const setCssVariables = ({ backgroundColor, highlightColor, headerBackground }: Styling) => {
  const root = document.querySelector(':root') as HTMLElement;

  if (root && backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }

  if (root && highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (root && headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }
};

const maybeInjectAnalyticsLibrary = (config: Config) => {
  if (!config.analyticsToken) return;

  return addScript('/jwpltx.js');
};

const calculateAccessModel = (config: Config): AccessModel => {
  const { id, monthlyOffer, yearlyOffer } = config?.integrations?.cleeng || {};

  if (!id) return 'AVOD';
  if (!monthlyOffer && !yearlyOffer) return 'AUTHVOD';
  return 'SVOD';
};

export const loadAndValidateConfig = async (
  onLoading: (isLoading: boolean) => void,
  onValidationError: (error: Error) => void,
  onValidationCompleted: (config: Config) => Promise<void>,
) => {
  onLoading(true);

  try {
    const configLocation = formatSourceLocation(getConfigLocation());

    if (!configLocation) {
      onValidationError(new Error('Config not defined'));
      return;
    }

    let config = await loadConfig(configLocation);

    if (!config) {
      return;
    }

    config = await validateConfig(config);
    config = merge({}, defaultConfig, config);

    // make sure the banner always defaults to the JWP banner when not defined in the config
    if (!config.assets.banner) {
      config.assets.banner = defaultConfig.assets.banner;
    }

    const accessModel = calculateAccessModel(config);

    useConfigStore.setState({
      config: config,
      accessModel,
    });

    setCssVariables(config.styling);
    maybeInjectAnalyticsLibrary(config);
    await onValidationCompleted(config);

    return config;
  } catch (ex: unknown) {
    onValidationError(ex as Error);
    return;
  }
};

function formatSourceLocation(source?: string) {
  if (!source) {
    return undefined;
  }

  if (source.match(/^[a-z,\d]{8}$/)) {
    return `${CONFIG_HOST}/apps/configs/${source}.json`;
  }

  return source;
}
