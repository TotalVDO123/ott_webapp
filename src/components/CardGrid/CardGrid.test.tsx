import * as React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import CardGrid from './CardGrid';

import { generatePlaylistPlaceholder } from '#src/utils/collection';

describe('<CardGrid>', () => {
  it('renders and matches snapshot', () => {
    const placeholderData = generatePlaylistPlaceholder();
    const { container } = render(
      <BrowserRouter>
        <CardGrid
          playlist={placeholderData}
          onCardHover={vi.fn()}
          onCardClick={vi.fn()}
          isLoading={false}
          accessModel={'SVOD'}
          isLoggedIn={true}
          hasSubscription={true}
        />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });
});
