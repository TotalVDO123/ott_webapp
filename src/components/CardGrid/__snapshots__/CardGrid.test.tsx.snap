// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CardGrid> > renders and matches snapshot 1`] = `
<div>
  <div>
    <div
      class="_container_7175d1 _cols-5_7175d1"
      role="grid"
    >
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_0/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_1/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_2/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_3/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_4/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_5/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_6/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_7/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_8/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_9/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_10/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_11/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_12/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_13/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <a
            aria-label=""
            class="_card_d75732"
            href="/m/placeholder_14/?r="
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
